src/python/pants/backend/python/util_rules/pex_from_targets.py

```diff
@@ -141,25 +139,28 @@ def __init__(
         :param description: A human-readable description to render in the dynamic UI when building
             the Pex.
         """
-        self.addresses = Addresses(addresses)
-        self.output_filename = output_filename
-        self.internal_only = internal_only
-        self.layout = layout
-        self.main = main
-        self.inject_args = tuple(inject_args)
-        self.inject_env = FrozenDict(inject_env)
-        self.platforms = platforms
-        self.complete_platforms = complete_platforms
-        self.additional_args = tuple(additional_args)
-        self.additional_lockfile_args = tuple(additional_lockfile_args)
-        self.include_source_files = include_source_files
-        self.include_requirements = include_requirements
-        self.include_local_dists = include_local_dists
-        self.additional_sources = additional_sources
-        self.additional_inputs = additional_inputs
-        self.hardcoded_interpreter_constraints = hardcoded_interpreter_constraints
-        self.description = description
+        object.__setattr__(self, "addresses", Addresses(addresses))
+        object.__setattr__(self, "output_filename", output_filename)
+        object.__setattr__(self, "internal_only", internal_only)
+        object.__setattr__(self, "layout", layout)
+        object.__setattr__(self, "main", main)
+        object.__setattr__(self, "inject_args", tuple(inject_args))
+        object.__setattr__(self, "inject_env", FrozenDict(inject_env))
+        object.__setattr__(self, "platforms", platforms)
+        object.__setattr__(self, "complete_platforms", complete_platforms)
+        object.__setattr__(self, "additional_args", tuple(additional_args))
+        object.__setattr__(self, "additional_lockfile_args", tuple(additional_lockfile_args))
+        object.__setattr__(self, "include_source_files", include_source_files)
+        object.__setattr__(self, "include_requirements", include_requirements)
+        object.__setattr__(self, "include_local_dists", include_local_dists)
+        object.__setattr__(self, "additional_sources", additional_sources)
+        object.__setattr__(self, "additional_inputs", additional_inputs)
+        object.__setattr__(
+            self, "hardcoded_interpreter_constraints", hardcoded_interpreter_constraints
+        )
+        object.__setattr__(self, "description", description)
```

Comment:
I think some of these could benefit from instead using a constructor method pattern instead..

```python
@dataclass(frozen=True)
class PexFromTargetsRequest:
  [field declarations...]

  @classmethod
  def create(cls, addresses: ...) -> PexFromTargetsRequest:
    return PexFromTargetsRequest(
      addresses=Addresses(addresses),
      output_filename=output_filename,
      ...
    )
 
  ...
```

This avoids the need for `object.__setattr__` at the cost of affecting every callsite to start using `PexFromTargetsRequest.create(...)` instead of just `PexFromTargetsRequest(...)`.

Maybe it could be worth investigating if we couldn't find a way to get both..? (i.e. validating/calculating the field values while still using the objects normal constructor)

This is in no way a comment blocking this particular change.. I'm just musing..

Reply:
If we didn't already have this interface for `__init__` that'd be a wonderful solution.

Unfortunately, I think we sailed that ship 😔

Reply:
yep.. for new classes tho ;)